feat: add root option to afterReset hooks of type command

jcgsville opened the pull request:
## Description

Fix for [issue 132](https://github.com/graphile/migrate/issues/132). afterReset actions of type command now respect the `root` property in the action spec

## Performance impact

I can't imagine a performance cost to this

## Security impact

None that I'm aware of

## Checklist

<!-- If this PR is work in progress, please open it as a "Draft PR". -->
<!-- To tick a checkbox, change it from `[ ]` to `[x]` -->

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [x] I've added tests for the new feature, and `yarn test` passes.
- [x] I have detailed the new feature in the relevant documentation.
- [ ] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [x] If this is a breaking change I've explained why.

I wasn't able to get my unit tests to run locally. Happy to try to write a test for this if you have any advice on what I'm missing? This was the behavior when I checked out the commit before mine so none of my changes are involved here.

```
yarn test
yarn run v1.22.10
$ yarn lint && depcheck --ignores @types/jest,eslint_d,tslib && yarn run test:only --ci
$ yarn prettier:check && eslint --ext .js,.jsx,.ts,.tsx,.graphql .
$ prettier --ignore-path .eslintignore --check '**/*.{js,jsx,ts,tsx,graphql,md,json}'
Checking formatting...
docs/docker/README.md
docs/idempotent-examples.md
README.md
Code style issues found in the above file(s). Forgot to run Prettier?
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Also, I don't see a RELEASE_NOTES.md, so I made no additions there.

benjie commented:
Huh! Looks like we weren't running prettier in CI?! Fixed in #138 - please rebase, then hopefully you can run the tests :)

benjie commented on src/actions.ts:
```diff
@@ -92,6 +92,9 @@ export async function executeActions(
         },
       );
     } else if (actionSpec._ === "command") {
+      const hookConnectionString = actionSpec.root
+        ? makeRootDatabaseConnectionString(parsedSettings, databaseName)
+        : connectionString;
```
Since you've added `root` to the `ActionSpecBase` let's move this to above the `if` statement so all the action types can use it easily.

benjie commented on src/actions.ts:
```diff
@@ -102,7 +105,7 @@ export async function executeActions(
           {
             GM_DBNAME: databaseName,
             GM_DBUSER: databaseUser,
```
This is meant to be "the database user in GM_DBURL", but since GM_DBURL might be a root connection string, this should then be the root user? But that wouldn't necessarily be expected... I'm worried this'll be misused... Maybe when `actionSpec.root` is used `GM_DBUSER` should not be set? I'm not sure people use it anyway.

Let's not set it when `root`. Please also update the README to reflect this.

jcgsville replied:
Ooh, good catch! Updating

benjie commented on README.md:
```diff
@@ -611,6 +611,9 @@ When the command is invoked it will have access to the following envvars:
 certainly mean to use `GM_DBURL` in your scripts since they will want to change
 whichever database was just reset/migrated/etc (which could be the shadow DB).
 
+The `root` property applies to `command` actions with the same affects as `sql`
+actions.
```
```suggestion
The `root` property applies to `command` actions with the same effects as `sql`
actions (see above).
```

benjie reviewed (changes requested):
I'm worried about unintended consequences of this, so we should probably be clear in the documentation exactly what this is expected to be used for.

jcgsville commented:
@benjie I think this PR is ready to go with the exception of the release notes checkbox. Not sure if I'm missing where to add those?

What's changed since your 👀 were on this:
1. I added more detail to the readme to make clear the behavior of the root property for command actions.
2. I modified GM_DBUSER to be undefined when root = true at your suggestion to avoid confusion. I also noted this in the readme.
3. I added a unit test testing that afterReset command actions with `root: true` behave as intended
4. I added a bonus unit test testing that afterReset SQL actions with `root: true` behave as intended

Apologies that I didn't realize force pushing to my branch would break Github's requested changes. The change you requested was a small tweak to the README which I implemented.

Finally, as I was working on the unit tests, I found a separate bug in `makeRootDatabaseConnectionString()`. This should probably be a separate PR to fix this. If it makes more since for it to be one, let me know.

The 🐛 : when setting rootConnectionString to just the db name, the connection string used when `root: true` incorrectly had a slash prepended to the database name.

For example, when `rootConnectionString: "postgres", connectionString: "postgres://localhost/app_db"` the connection string when `root: true` should be `postgres` but it was `/postgres`.

I noticed this because the unit tests default rootConnection string was just `postgres` and the assertions were failing unexpectedly. I've modified `settings.ts:420` to fix it, and I believe this change produces the desired behavior. But it's worth your 👀 to make sure I didn't misunderstand something. I also added a unit test to cover this case.

jcgsville commented:
@benjie know of anything left before this can be merged?

benjie reviewed (changes requested):
Great work, some small tweaks needed :+1:

jcgsville commented on __tests__/actions.test.ts:
```diff
@@ -74,6 +79,42 @@ it("runs command actions", async () => {
+  mockedWithClient.mockClear();
+  await executeActions(parsedSettings, false, parsedSettings.afterReset);
+  expect(mockedWithClient).toHaveBeenCalledTimes(1);
+  expect(mockedWithClient.mock.calls[0][0]).toBe("graphile_migrate_test");
```
Ah, good catch. I missed this completely.

jcgsville commented on __tests__/actions.test.ts:
```diff
@@ -74,6 +79,42 @@ it("runs command actions", async () => {
+  );
+
+  await executeActions(parsedSettings, false, parsedSettings.afterReset);
+  expect(mockedExec.mock.calls[0][1].env.GM_DBUSER).toBe(undefined);
```
Good call. Very difficult to understand what's going on here without this PR's discussion as context 👍

jcgsville replied:
actions.ts as well

jcgsville commented on __tests__/actions.test.ts:
```diff
@@ -74,6 +79,42 @@ it("runs command actions", async () => {
+  await executeActions(parsedSettings, false, parsedSettings.afterReset);
+  expect(mockedExec.mock.calls[0][1].env.GM_DBUSER).toBe(undefined);
+  expect(mockedExec.mock.calls[0][1].env.GM_DBURL).toBe(
+    "graphile_migrate_test",
```
Agreed, that would be much better. Added assertions for most parts of the URL.

jcgsville commented on src/actions.ts:
```diff
@@ -101,8 +101,8 @@ export async function executeActions(
             GM_DBNAME: databaseName,
-            GM_DBUSER: databaseUser,
-            GM_DBURL: connectionString,
+            GM_DBUSER: actionSpec.root ? undefined : databaseUser,
```
Ah, the joys of JS undefined. Updated 👍

jcgsville commented on src/settings.ts:
```diff
@@ -417,7 +417,7 @@ export function makeRootDatabaseConnectionString(
     }
   } else {
-    parsed.pathname = `/${databaseName}`;
+    parsed.pathname = databaseName;
```
My first gut reaction was that I'd rather test it thoroughly at dev time than introduce a branch at runtime, but the docs on url show the pathname going in and coming out always with a leading slash. The docs make no reference to the leading slash's behavior as a part of the API, so, I think you're right here, that it would be risky.

https://nodejs.org/api/url.html#urlpathname

+ I also realized that since url is a core library, the behavior of url is not encoded in package.json / the lock file, so the dependency could change out from under us without a CI run if it's broken in a later node version.

Interestingly, url's `parse()` treats the string "somedb" as, in fact, a path. I'm not familiar enough with the formal URL spec here to know if the library is behaving according to spec, but I'm assuming it is. It sounds like we can't rely on the case where the `url` lib thinks of the string as a non-url.

```
Welcome to Node.js v16.13.2.
Type ".help" for more information.
> require('url').parse('somedb');
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: 'somedb',
  path: 'somedb',
  href: 'somedb'
}
> require('url').format(require('url').parse('somedb'));
'somedb'
> require('url').format(require('url').parse('/somedb'));
'/somedb'
```

jcgsville replied:
Apologies for the word vomit.

I read through the URI RFC, and I thought it made sense to trust that the url library will reliably use the path as a delimiter between the host component and path component of the URI. If it didn't, it would be out-of-spec-ish with the RFC.

[Section 1.2.3](https://www.rfc-editor.org/rfc/rfc3986#section-1.2.3) says that `For some URI schemes, the visible hierarchy is limited to the scheme itself: everything after the scheme component delimiter (":") is considered opaque to URI processing.  Other URI schemes make the hierarchy explicit and visible to generic parsing algorithms. The generic syntax uses the slash ("/"), question mark ("?"), and number sign ("#") characters to delimit components that are significant to the generic parser's hierarchical interpretation of an identifier`

This certainly implies that the slash aught to be used as a delimiter, but it doesn't explicitly say it must.

[Section 3.2](https://www.rfc-editor.org/rfc/rfc3986#section-3.2) says `If a URI contains an authority component, then the path component must either be empty or begin with a slash ("/") character` this is the important line. This suggests to me that if the format function did not prepend a slash on a path when there is an authority of any kind, it would be violating the spec.

So this suggests we're good to rely on the slash, and yet...

```
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: 'postgres' });
'postgres:abc.comfoobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: null, auth: 'user:pass' });
'user:pass@abc.comfoobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: 'postgres', slashes: true });
'postgres://abc.com/foobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: null });
'abc.comfoobar'
> require('url').format({ pathname: 'foobar', hostname: 'abc.com', protocol: null, slashes: true });
'//abc.com/foobar'
```

I did test that when `slashes: true`, it will always use a slash as delimiter between hostname and path, and that if the url contains `//` between the scheme and the hostname, then calling parse() will always return a url object with `slashes: true`.  This specific behavior of parse is unfortunately undocumented, and thus we fall into the same risk of it being susceptible to unexpected breaking changes.

The behavior above is only out-of-spec-ish because the URI RFC basically says "meh, schemes can do kinda whatever they want. They don't necessarily have to respect the generic parsing algorithms"

So... let's go see what postgres has to say about this. See `33.1.1.2. Connection URIs` in [docs](https://www.postgresql.org/docs/12/libpq-connect.html). It specifically says that connection URIs must start with `postgres` or `postgresql`, and it contains no documentation suggesting that "connection strings" can come in the form of a standalone db name.

[Node-postgres](https://node-postgres.com/features/connecting) specifically references Connection URIs. The docs do not suggest that `somedbname` is a valid connection uri. But if I try it... it works

```
> const { Pool } = require('pg')
undefined
> new Pool({ connectionString: 'template1' }).query('select now()', console.log);
undefined
> undefined Result {
  command: 'SELECT',
  rowCount: 1,
  oid: null,
  rows: [ { now: 2022-03-01T07:11:29.109Z } ],
  ...
}
```

I'm guessing this is just a psql convention that node-postgres adopted. [Psql docs](https://www.postgresql.org/docs/13/app-psql.html). `If an argument is found that does not belong to any option it will be interpreted as the database name`

I'm assuming graphile migrate would want to follow node-postgres's lead, even if it's not well documented. If so, I'd recommend adding a regex that checks if it's a valid db name, and if so, treat it differently.

Unfortunately... there are no restrictions for postgres database names beyond those of identifiers, which means you can do a lot including unicode characters 😳 

```
create database U&"\0441\043B\043E\043D";
CREATE DATABASE
Time: 0.177s
```

This works!

```
> new Pool({ connectionString: '\u0441\u043B\u043E\u043D' }).query('select now()', console.log);
undefined
> undefined Result {
  command: 'SELECT',
  rowCount: 1,
  oid: null,
  rows: [ { now: 2022-03-01T07:23:42.934Z } ],
  ...
}
```

Do we really want to go down this path?

(Wow the 🐰 🕳️ is real when trying to protect against people using a library in unexpected ways)

benjie replied:
Can we just gate it on `parsed.protocol`?

jcgsville replied:
... yep. That seems like that will work 😅 

Updated function

jcgsville commented:
I addressed the recommendations to fix the tests. Last thing pending is resolution surrounding the pathname vs dbname debacle

benjie commented on __tests__/actions.test.ts:
```diff
@@ -74,6 +81,48 @@ it("runs command actions", async () => {
+  // When `root: true`, GM_DBUSER may be perceived as ambiguous, so we must not set it.
+  expect(mockedExec.mock.calls[0][1].env.GM_DBUSER).toBe(undefined);
+  const connectionStringParts = parse(TEST_DATABASE_URL);
+  const rootConnectionStringParts = parse(TEST_ROOT_DATABASE_URL);
```
Just to make the test clearer, we should assert that the databases were previously different.

```suggestion
  const rootConnectionStringParts = parse(TEST_ROOT_DATABASE_URL);
  expect(rootConnectionStringParts.database).not.toBe(connectionStringParts.database);
```

benjie reviewed (changes requested):
Looks good to me! :+1:

benjie commented:
My new test didn't pass linting :(

benjie commented:
I tried fixing it using prettier playground but that wasn't enough; please reformat :+1:

jcgsville commented:
Thanks for your thorough reviews!

benjie commented:
Of course it was the trailing comma :man_facepalming:

benjie commented:
Released in 1.3.0 :raised_hands:
